gateway/src/apicast/policy/3scale_batcher/reports_batcher.lua

```diff
@@ -0,0 +1,90 @@
+local keys_helper = require('apicast.policy.3scale_batcher.keys_helper')
+
+local setmetatable = setmetatable
+local ipairs = ipairs
+local insert = table.insert
+local resty_lock = require 'resty.lock'
+
+local _M = {}
+
+local mt = { __index = _M }
+
+local lock_timeout = 10
+
+-- Note: storage needs to implement shdict interface.
+function _M.new(storage, lock_shdict_name)
+  local self = setmetatable({}, mt)
+  self.storage = storage
+  self.lock_shdict_name = lock_shdict_name
+  return self
+end
+
+-- Note: When the lock cannot be acquired, the report will be lost.
+-- The timeout is high, that means that if it could not be acquired, there's
+-- probably a problem in the system.
+-- TODO: Find a solution for this.
+function _M:add(service_id, credentials, usage)
+  local deltas = usage.deltas
+
+  local lock, new_lock_err = resty_lock:new(self.lock_shdict_name, { timeout = lock_timeout })
```

Comment:
The options with the timeout can be pre-allocated so it does not need to be allocated on every call.

Reply:
Maybe even the lock object can be allocated in .new and stored in self ?

Reply:
If I understood correctly, that's not recommended: https://github.com/openresty/lua-resty-lock#for-multiple-lua-light-threads

Reply:
Yes. Fixed the options part :+1: